# Part 1. Equation of a Slime

How many late days are you using for this assignment? - 0

In [25]:
# Imports section
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [26]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
science_dataset = pd.read_csv('science_data_large.csv')
display(science_dataset.head(15))
display(science_dataset.describe())
science_dataset.info()

,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


,Temperature °C,Mols KCL,Size nm^3
count,1000.000000,1000.000000,1.000000e+03
mean,500.500000,471.530000,5.086111e+05
std,288.819436,288.482872,4.474838e+05
min,1.000000,1.000000,1.611429e+01
25%,250.750000,226.750000,1.298267e+05
50%,500.500000,459.500000,3.827182e+05
75%,750.250000,710.250000,7.603211e+05
max,1000.000000,1000.000000,1.972127e+06


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


## 2. Splitting the dataset

In [27]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = science_dataset[['Temperature °C', 'Mols KCL']]  # Features
y = science_dataset['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [28]:
# Use sklearn to train a model on the training set
lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_datapoint = pd.DataFrame({'Temperature °C': [123], 'Mols KCL': [683]})
predicted_output = lr_model.predict(sample_datapoint)
print("Predicted Output:", predicted_output)


# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
model_score = lr_model.score(X_test, y_test)
print("Model Score:", model_score)


# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = lr_model.coef_
print("Coefficients:", coefficients)
intercept = lr_model.intercept_
print("Intercept:", intercept)

Predicted Output: [402475.25967114]
Model Score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340775


Write the linear equation of a slime: (example equation: $E = mc^2$)

$T = \text{Temperature °C}$\
$M = \text{Mols KCL}$

$$h(X) = 866.14641 \cdot \text{T} + 1032.69507 \cdot \text{M} - 409391.47958$$

Report on score and explain meaning:


In simple terms, the score represents the proportion of the variance in the dependent variable (in this case, the change in size of the slime) that can be explained by the independent variables (temperature and KCl concentration). A score of 1.0 would mean that the model perfectly predicts the output with no error, while a score less than that would indicate errors in the model's prediction.

The score of the linear regression model, which is approximately 0.855, indicates the goodness of fit for the model when predicting the growth of the slime based on temperature and potassium chloride (KCl) concentration. In our case, an $R^2$ score of 0.855 indicates that approximately 85.5% of the variability in the size of the slime can be explained by the linear combination of temperature and potassium chloride concentration used in our model. This high score suggests a strong correlation between the inputs and the output, implying that the model fits the data well, leaving only about 14.5% of the variance unexplained.

Overall, this score reassures us that our model is an effective tool for predicting the behavior of this new life form under the tested conditions.

## 4. Use Cross Validation

In [29]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

cv = KFold(n_splits=5, random_state=42, shuffle=True)
scores = cross_val_score(lr_model, X, y, cv=cv)
print("Scores:", scores)
mean = np.mean(scores)
print("Mean:", mean)

# Report on their finding and their significance

Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean: 0.8597294202684646


Write findings here:

The cross-validation results yielded scores of approximately [0.861, 0.827, 0.872, 0.882, 0.856], with a mean score of 0.860.

In simple terms, the mean score represents that about 86.0% of the variability in the slime’s size can be explained by temperature and potassium chloride concentration. Each score reflects how well the model predicts slime growth across different data subsets, showing robust performance as all values exceed 0.82.

This consistent high performance reassures us that the linear regression model effectively captures the relationship between the inputs and slime growth. Furthermore, the slight variability in scores suggests potential areas for model improvement through feature exploration or alternative modeling approaches.

Overall, these findings corroborate the result from part 3 and the reliability of our model.

## 5. Using Polynomial Regression

In [30]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)
print("Columns:", poly.get_feature_names_out())

X_train, X_test = train_test_split(X_poly, test_size=0.1, random_state=42)

lr_model_poly = LinearRegression(fit_intercept=False)
lr_model_poly.fit(X_train, y_train)

sample_datapoint = poly.fit_transform(pd.DataFrame({'Temperature °C': [123], 'Mols KCL': [683]}))
predicted_output = lr_model_poly.predict(sample_datapoint)
print("Predicted Output:", predicted_output)

model_score = lr_model_poly.score(X_train, y_train)
print("Model Score:", model_score)

coefficients = lr_model_poly.coef_
print("Coefficients:", coefficients)
intercept = lr_model_poly.intercept_
print("Intercept:", intercept)

cv = KFold(n_splits=5, random_state=42, shuffle=True)
scores = cross_val_score(lr_model_poly, X_poly, y, cv=cv)
print("Cross Validation Scores:", scores)
mean = np.mean(scores)
print("Cross Validation Mean:", mean)

# Report on the metrics and output the resultant equation as you did in Part 3.

Columns: ['1' 'Temperature °C' 'Mols KCL' 'Temperature °C^2'
 'Temperature °C Mols KCL' 'Mols KCL^2']
Predicted Output: [182822.25713936]
Model Score: 1.0
Coefficients: [ 2.04775586e-05  1.20000000e+01 -1.27196139e-07  1.26458843e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 0.0
Cross Validation Scores: [1. 1. 1. 1. 1.]
Cross Validation Mean: 1.0


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$T = \text{Temperature °C}$\
$M = \text{Mols KCL}$

$h(x) = 2.0478 \cdot 10^{-05} + 12 \cdot T - 1.2720 \cdot 10^{-07}\cdot M + 1.2646 \cdot 10^{-11}\cdot T^2 + 2 \cdot TM + 2.8571 \cdot 10^{-02} \cdot M^2$

$$\Rightarrow h(x) = 2 \cdot 10^{-05} + 12 \cdot T + 2 \cdot TM + 2.857 \cdot 10^{-02} \cdot M^2$$

Report on the score and interpret:

The polynomial regression model yielded a perfect score of 1.0, indicating that it perfectly predicts the output based on the provided independent variables (temperature and potassium chloride concentration). In simpler terms, this means that 100% of the variance in the change in size of the slime can be explained by the model, leaving no error in the predictions.

The perfect score of $R^2 = 1.0$ indicates that the polynomial regression model captures the relationship between the independent variables and slime growth perfectly. This model may provide critical insights into the growth dynamics of the slime, potentially leading to significant discoveries in our understanding of this extraterrestrial life form.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [31]:
# Load the dataset. Then train and evaluate the classification models.
ckd_dataset = pd.read_csv('ckd_feature_subset.csv')
display(ckd_dataset.head(15))
ckd_dataset.info()

,age,bp,wbcc,appet_poor,appet_good,rbcc,Target_ckd
0,0.688312,0.333333,0.000000,1,0,0.000000,1
1,0.545455,0.333333,0.128319,1,0,0.305085,1
2,0.714286,0.500000,0.238938,1,0,0.186441,1
3,0.688312,0.333333,0.283186,0,1,0.338983,1
4,0.441558,0.333333,0.221239,1,0,0.220339,1
5,0.740260,0.833333,0.265487,0,1,0.355932,1
6,0.870130,0.333333,0.668142,0,1,0.237288,1
7,0.870130,0.500000,0.150442,0,1,0.372881,1
8,0.194805,0.666667,0.380531,0,1,0.305085,1
9,0.753247,0.833333,0.163717,0,1,0.220339,1


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   age         153 non-null    float64
 1   bp          153 non-null    float64
 2   wbcc        153 non-null    float64
 3   appet_poor  153 non-null    int64  
 4   appet_good  153 non-null    int64  
 5   rbcc        153 non-null    float64
 6   Target_ckd  153 non-null    int64  
dtypes: float64(4), int64(3)
memory usage: 8.5 KB


In [32]:
X = ckd_dataset.drop('Target_ckd', axis=1)  # Adjust 'class' if your target column is named differently
y = ckd_dataset['Target_ckd'] 

models = {
    "Logistic Regression": LogisticRegression(random_state=42),
    "Support Vector Machine": SVC(random_state=42),
    "k-Nearest Neighbors": KNeighborsClassifier(),
    "Neural Network": MLPClassifier(max_iter=1500, random_state=42)
}

# Perform 5-fold cross validation
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Store the results
results = {}

for name, model in models.items():
    scores = cross_val_score(model, X, y, cv=kf)
    results[name] = {
        'Mean Accuracy': np.mean(scores),
        'Standard Deviation': np.std(scores)
    }

# Convert results to a DataFrame for better visualization
results_df = pd.DataFrame(results).T
results_df

,Mean Accuracy,Standard Deviation
Logistic Regression,0.856559,0.066269
Support Vector Machine,0.928172,0.047601
k-Nearest Neighbors,0.927957,0.052440
Neural Network,0.935054,0.040466


## Results and Conclusion for Classification Experiments

### Model Performance Summary

| Model                  | Mean Accuracy | Standard Deviation |
|-----------------------|---------------|---------------------|
| Logistic Regression    | 0.856559      | 0.066269            |
| Support Vector Machine | 0.928172      | 0.047601            |
| k-Nearest Neighbors    | 0.927957      | 0.052440            |
| Neural Network         | 0.935054      | 0.040466            |

The logistic regression model had the worst accuracy of 0.856559 and the highest standard deviation of 0.066269. This means it had the worst result of all the models and that its results were less consistent than other models.

The SVM performed significantly better than the Logistic Regression and more in line with the other models with an accuracy of 0.928172 and a standard deviation of 0.047601. This means it is both very accurate and also consistent with its results.

The kNN algorithm was marginally less accurate than the SVM with a mean of 0.927957 and it had a slightly higher standard deviation of 0.052440. This means that the model's accuracy is very similar to the SVM but is is less consistent than the SVM algorithm.

Finally, the neural network was the most accurate with a mean accuracy of 0.927957 and the most consistent with a standard deviation of 0.040466.

The neural network is the best performing model in this experiment, achieving the highest mean accuracy with the lowest standard deviation. This suggests that it not only correctly classifies a high percentage of instances but also maintains consistency across the different subsets of the dataset. The results suggest that the chronic kidney disease classification task benefits from models that can capture the underlying complexity in the data. The neural network's  performance implies it can model intricate relationships between features, which may be critical for accurately predicting CKD.

In [33]:
# Experiments with Neural Network.
configurations = [
    {'hidden_layer_sizes': (10), 'max_iter': 100},
    {'hidden_layer_sizes': (10), 'max_iter': 1000},
    {'hidden_layer_sizes': (50), 'max_iter': 100},
    {'hidden_layer_sizes': (50), 'max_iter': 1000}
]

# Store results for each configuration
nn_results = {}

for config in configurations:
    model = MLPClassifier(hidden_layer_sizes=config['hidden_layer_sizes'], activation='relu', max_iter=config['max_iter'], random_state=42)
    scores = cross_val_score(model, X, y, cv=kf)
    nn_results[f"Layer Size: {config['hidden_layer_sizes']}, Max Iterations: {config['max_iter']}"] = {
        'Mean Accuracy': np.mean(scores),
        'Standard Deviation': np.std(scores)
    }

# Convert results to a DataFrame for better visualization
nn_results_df = pd.DataFrame(nn_results).T
nn_results_df

/home/arstgi/miniconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/arstgi/miniconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/arstgi/miniconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/arstgi/miniconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/arstgi/miniconda3/lib/pyth

,Mean Accuracy,Standard Deviation
"Layer Size: 10, Max Iterations: 100",0.824086,0.074418
"Layer Size: 10, Max Iterations: 1000",0.908602,0.052041
"Layer Size: 50, Max Iterations: 100",0.843441,0.065366
"Layer Size: 50, Max Iterations: 1000",0.935054,0.040466


## Results and Conclusion for Neural Network Experiments

### Model Performance Summary
| Configuration | Mean Accuracy | Standard Deviation |
|-------------------------------------------|---------------|---------------------|
| Layer Size: 10, Max Iterations: 100 | 0.824086 | 0.074418 |
| Layer Size: 10, Max Iterations: 1000 | 0.908602 | 0.052041 |
| Layer Size: 50, Max Iterations: 100 | 0.843441 | 0.065366 |
| Layer Size: 50, Max Iterations: 1000 | 0.935054 | 0.040466 |


Results Analysis
1. Layer Size: 10, Max Iterations: 100
    - This configuration yielded the lowest mean accuracy of 82.4% with a standard deviation of 7.4%. The model's performance suggests that a smaller hidden layer with fewer iterations may not capture the complexities of the data adequately.
2. Layer Size: 10, Max Iterations: 1000
    - Increasing the iterations to 1000 improved accuracy significantly to 90.9%, with a standard deviation of 5.2%. This indicates that allowing more training time leads to better learning from the data, even with a small hidden layer.
3. Layer Size: 50, Max Iterations: 100
    - With a hidden layer size of 50 and 100 iterations, the accuracy slightly improved to 84.3% with a standard deviation of 6.5%. This suggests that increasing the layer size helps but still requires more iterations for optimal performance.
4. Layer Size: 50, Max Iterations: 1000
    - The best performance came from this configuration, achieving a mean accuracy of 93.5% and a standard deviation of 4.0%. This result indicates that both a larger hidden layer and higher iteration count work synergistically to capture complex patterns in the data effectively.

Looking at all these results, we can say that increasing the number of iterations has a greater impact than increasing the hidden layer size, especially when the number of iterations is below the limit for the model to converge. We might be able to get better results, after increasing the number of iterations till it converges, by both increasing the hidden layer size and the number of hidden layers beyond one.